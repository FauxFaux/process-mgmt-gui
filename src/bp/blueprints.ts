// https://github.com/FauxFaux/factorio-loader/blob/98eb701ed55afb0dc01d40f5b05fb63921fa2eb3/web/muffler/blueprints.ts

import { atob, btoa } from 'abab';
import pako from 'pako';

export interface Blueprint {
  entities?: Entity[];
  tiles?: { name: string; position?: unknown }[];

  icons?: unknown[];

  // from the map version
  version: number;
  label?: string;
  item: 'blueprint';

  // incomplete
}

export interface Entity {
  entity_number?: number;

  // presumably always an item; this is *not* colon'd
  name: string;
  inventory?: unknown;
  // e.g. { speed_module: 6 } for factories, not an inventory apparently
  items?: Record<string, number>;
  position?: unknown;
  recipe?: string;
  direction?: number;
  // massively over-specified
  control_behavior?: {
    logistic_condition: {
      first_signal: {
        type: 'item';
        name: string;
      };
      constant: number;
      comparator: '<';
    };
    connect_to_logistic_network: true;
  };
  // e.g. power poles
  neighbours?: number[];

  request_filters?: { name: string; count: number; index?: number }[];
}

export function decode(data: string): Blueprint {
  if (data[0] !== '0') throw new Error(`unsupported version ${data[0]}`);
  const un64 = Uint8Array.from(atob(data.slice(1))!, (c) => c.charCodeAt(0));
  const obj = JSON.parse(pako.inflate(un64, { to: 'string' }));
  const keys = Object.keys(obj);
  if (keys.length !== 1 || keys[0] !== 'blueprint')
    throw new Error(`invalid top level: ${keys}`);
  return obj.blueprint as Blueprint;
}

export function encode(blueprint: Blueprint): string {
  const obj = {
    blueprint,
  };
  const un64 = pako.deflate(JSON.stringify(obj), { level: 9 });
  return '0' + btoa(String.fromCharCode(...un64));
}
