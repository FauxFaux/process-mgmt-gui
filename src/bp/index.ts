import type { Blueprint } from './blueprints';

// c.f. https://github.com/FauxFaux/factorio-loader/blob/aca1e2a28c69e0ac073391b6498f897d6803d02a/web/pages/recipes.tsx#L247
// c.f. https://github.com/FauxFaux/factorio-loader/blob/aca1e2a28c69e0ac073391b6498f897d6803d02a/web/pages/recipes.tsx#L299
export const mallAssembler = (
  recipeId: string,
  itemId: string,
  itemHumanName: string | undefined,
  inputs: Record<string, number>,
): Blueprint => ({
  entities: [
    {
      entity_number: 1,
      name: 'assembling-machine-2',
      position: {
        x: -124.5,
        y: 130.5,
      },
      recipe: recipeId,
    },
    {
      control_behavior: {
        connect_to_logistic_network: true,
        logistic_condition: {
          comparator: '<',
          constant: 5,
          first_signal: {
            name: itemId,
            type: 'item',
          },
        },
      },
      direction: 6,
      entity_number: 2,
      name: 'fast-inserter',
      position: {
        x: -122.5,
        y: 130.5,
      },
    },
    {
      direction: 2,
      entity_number: 3,
      name: 'fast-inserter',
      position: {
        x: -122.5,
        y: 131.5,
      },
    },
    {
      entity_number: 4,
      name: 'logistic-chest-passive-provider',
      position: {
        x: -121.5,
        y: 130.5,
      },
    },
    {
      entity_number: 5,
      name: 'logistic-chest-requester',
      position: {
        x: -121.5,
        y: 131.5,
      },
      request_filters: Object.entries(inputs).map(([name, count], i) => ({
        count,
        index: i + 1,
        name,
      })),
    },
  ],
  icons: [
    {
      index: 1,
      signal: {
        name: 'assembling-machine-2',
        type: 'item',
      },
    },
    {
      index: 2,
      signal: {
        name: itemId,
        type: 'item',
      },
    },
  ],
  item: 'blueprint',
  label: 'mall' + (itemHumanName ? `: ${itemHumanName}` : ''),
  version: 281479277838336,
});
