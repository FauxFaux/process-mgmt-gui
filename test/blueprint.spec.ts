import { describe, expect, it } from '@jest/globals';
import { decode } from '../src/bp/blueprints';

describe('blueprint', () => {
  it('decodes examples', () => {
    const seTemplate = decode(
      '0eNqlU21vmzAQ/ivVfYZqmCRkaP9kqpCBS3KasZl9sEUR/33n0KRJS9ep+4SM73m5584nqM2AvSfLUJ6AGmcDlN9PEGhvtYn/+NgjlECMHSRgdRdPOgTsakN2n3a6OZDFVMGUANkWf0OZTcmHFMbtKTA1aXPAwKlumEZMe+9GatHfkKnpKQG0TEw4mzsfjpUduloqy+zvthLoXRCws9GLEKaZWj2uEzgKNP/yuBYpjw31/+Tqjbq6qu+0IMgG9CwXS7LqlWxLIjyXbBKQ8Nk7U9V40CM5H3EXP5Vctle2HfnA1X8HHKfNOo4+K+Kp67XXHIXh2/O9FX8Vu+rqwyL/cv4HlOwHnBbyyD+VR/Y6D7VAvXqvt14Gf9fcklh2F/4C/fo9eo8/B/l+yJs979K5utqREci8sZd3cVVAI416Z6MG+WYghpj3EIexKaK5y/a/LHc7attg+xag1A3gJX9xjGZu4a72KT6o866UN88/AaNrlHWCThtTPsz78nAT6SjdzLPZZqviqyqKbb7N8800/QGdqnC/',
    );
    expect(seTemplate).toMatchSnapshot();
  });
});
