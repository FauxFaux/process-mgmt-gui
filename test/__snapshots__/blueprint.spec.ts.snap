// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blueprint decodes examples 1`] = `
{
  "entities": [
    {
      "entity_number": 1,
      "name": "assembling-machine-2",
      "position": {
        "x": -124.5,
        "y": 130.5,
      },
      "recipe": "logistic-chest-active-provider",
    },
    {
      "control_behavior": {
        "connect_to_logistic_network": true,
        "logistic_condition": {
          "comparator": "<",
          "constant": 17,
          "first_signal": {
            "name": "logistic-chest-active-provider",
            "type": "item",
          },
        },
      },
      "direction": 6,
      "entity_number": 2,
      "name": "fast-inserter",
      "position": {
        "x": -122.5,
        "y": 130.5,
      },
    },
    {
      "direction": 2,
      "entity_number": 3,
      "name": "fast-inserter",
      "position": {
        "x": -122.5,
        "y": 131.5,
      },
    },
    {
      "entity_number": 4,
      "name": "logistic-chest-passive-provider",
      "position": {
        "x": -121.5,
        "y": 130.5,
      },
    },
    {
      "entity_number": 5,
      "name": "logistic-chest-requester",
      "position": {
        "x": -121.5,
        "y": 131.5,
      },
      "request_filters": [
        {
          "count": 67,
          "index": 1,
          "name": "electronic-circuit",
        },
        {
          "count": 22,
          "index": 2,
          "name": "advanced-circuit",
        },
        {
          "count": 22,
          "index": 3,
          "name": "steel-chest",
        },
      ],
    },
  ],
  "icons": [
    {
      "index": 1,
      "signal": {
        "name": "assembling-machine-2",
        "type": "item",
      },
    },
    {
      "index": 2,
      "signal": {
        "name": "logistic-chest-active-provider",
        "type": "item",
      },
    },
  ],
  "item": "blueprint",
  "label": "mall: active provider",
  "version": 281479277838336,
}
`;
